### PR TITLE
Editor page

### LINKED ISSUE #29: draw.io embed page
Features
- [ ] Easy to find UML diagram components
- [ ] Ability to save diagrams to the DB
- [ ] Ability to cache diagrams locally
- [ ] Autosave locally and to the DB
- [ ] Name a diagram
- [ ] Load a diagram from local storage or DB

### PR DESCRIPTION
**Pull request by Vladmidir**
# Summary

## General
Implemented the draw.io embed in the Editor page. 

The user is allowed to have one diagram stored in the db for each challenge. The diagram is autosaved every 10 seconds. The diagram is auto loaded for the specific challenge when clicking `Open Editor` button.

This is the initial pr. There are several TODOs left, including:
- Configure the editor styles to better suit UML. ✅
- Give descriptive name to the diagram ✅
- Integrate the editor with editing diagrams for existing solutions. `separate pr`
- Redo the server side logic to work with Eren's implementation of diagram file storage. `separate pr`

## Important 
- I have added a middleware function to mock authentication. I will have to modify it to match the shibboleth auth. That is something we will need for development, as we are not developing on the uoft's vm, and don't have the access to the actual shibboleth in development.

# Screenshot
image

close #29

### CODE REVIEW

File: client/src/pages/Challenge.tsx [hunk 1]
```diff
@@ -8,8 +8,9 @@ import { StarFill } from "react-bootstrap-icons";
 
 
 const Challenge = () => {
-    //TODO: Fetch this value from the server
-    const [completed, setCompleted] = useState(false); //TODO: This value should be fetched from the server
+    //TODO: Fetch this value from the server. 
+    // We can query the db for solutions with the current user id and the challenge id
```
Comment (alexapostolu) [inline client/src/pages/Challenge.tsx:12]: Good use of comments

File: client/src/pages/Challenge.tsx [hunk 2]
```diff
@@ -92,23 +102,17 @@ const Challenge = () => {
             </Col>
             :
             <Col>
-                <ButtonToolbar className="d-flex align-items-end justify-content-around">
-                    <Button className="m-1" target="_blank" href="/editor">Open Editor</Button>
-                    <Button className="m-1" href={"/solutions/post/" + id}>Post a Solution</Button>
-                    <Button className="m-1" target="_blank" href={"/solutions/challenge/" + id}>View Solutions</Button>
-                </ButtonToolbar>
+                <ChallengeButtons />
```
Comment (alexapostolu) [inline client/src/pages/Challenge.tsx:105]: Nice use of components

File: server/controllers/UserController.js [hunk 3]
```diff
@@ -42,7 +42,8 @@ exports.getComments = async (req, res) => {
 
 exports.create = async (req, res) => {
     try {
-        const { username, passwordHash, preferredName, email, score } = req.body;
+        const { username, passwordHash, preferredName, email } = req.body;
+        const score = 0;
```
Comment (alexapostolu) [inline server/controllers/UserController.js:46]: Why did you separate score?

File: server/models/SolutionInProgress.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
+module.exports = (sequelize, DataTypes) => {
+
+//TODO: add proper spacing and indentation
+const SolutionInProgress = sequelize.define('SolutionInProgress', {
```
Comment (alexapostolu) [inline server/models/SolutionInProgress.js:4]: Watch the indentation

**Review by alexapostolu — COMMENTED** [review]
Very good commenting skills Vlad, just watch the indentation a little bit

File: server/controllers/UserController.js (hunk 3 above)
Comment (Vladmidir) [reply server/controllers/UserController.js]: Passing score is redundant. A new user will have a score of zero. Always.

File: client/src/pages/Editor.tsx [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+                    onClose={(data) => {
+                        // Save and exit. Timeout to make sure the save request is sent before the tab is closed.
+                        doExport();
+                        setTimeout(() => {
+                            window.close();
+                        }, 1000);
+                    }}
+                    />
+            </div>
+            {/* TESTING ONLY
```
Comment (arthurhohinng) [inline client/src/pages/Editor.tsx:220]: Is it intentional to keep this commented block?

File: client/src/pages/Editor.tsx [hunk 6]
```diff
@@ -0,0 +1,236 @@
+import { useEffect, useRef, useMemo, useState, useCallback } from 'react';
+import { Button } from 'react-bootstrap';
+import { ActionExport, DrawIoEmbed, DrawIoEmbedRef} from 'react-drawio';
+import { useLocation } from 'react-router-dom';
+// import { redirect } from 'react-router-dom';
```
Comment (arthurhohinng) [inline client/src/pages/Editor.tsx:5]: Remove if not using

File: client/src/pages/Editor.tsx [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+
+
+const CONFIG = {
+    defaultEdgeStyle: {endSize: 12, startSize: 12},
+    // autosaveDelay: 10000, // 10 seconds
```
Comment (arthurhohinng) [inline client/src/pages/Editor.tsx:10]: Intentional to have autosave commented as well?

File: client/src/pages/Editor.tsx [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+    }
+
+    const query = useQuery();
+
+
+    //autosave (export) the diagram every 10 seconds (ONLY FOR DIAGRAMS IN PROGRESS, NOT EDITING EXISTING SOLUTIONS)
+    useEffect(() => {
+        if (query.get("type") !== "challenge") return;
+        const interval = setInterval(() => {
+            // console.log("autosaving");
```
Comment (arthurhohinng) [inline client/src/pages/Editor.tsx:48]: Commented for testing?

File: client/src/pages/Editor.tsx [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+                }
+                return response.json() as Promise<{diagram: string, title: string, id: string}>
+            })
+            .then(data => {
+                // console.log(data);
+                diagramData.current = data.diagram;
+                diagramId.current = data.id;
+                setDiagramName(data.title);
+                diagramNameRef.current = data.title; //Crutch for the handleSave function
+                // console.log("Loaded diagram: " + diagramData.current + " with id: " + diagramId.current + " and name: " + diagramNameRef.current);
```
Comment (arthurhohinng) [inline client/src/pages/Editor.tsx:78]: Same here

File: client/src/pages/Editor.tsx [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+                if(response.status === 404) {
+                    throw new Error("No diagram found for this challenge");
+                }
+                if (!response.ok) {
+                    throw new Error(response.statusText);
+                }
+                return response.json() as Promise<{diagram: string, title: string, id: string}>
+            })
+            .then(data => {
+                // console.log(data);
```
Comment (arthurhohinng) [inline client/src/pages/Editor.tsx:73]: And here

File: client/src/pages/Editor.tsx [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+                        if (!response.ok) {
+                            throw new Error(response.statusText);
+                        }
+                        return response.json() as Promise<{id: string, title: string}>
+                    }).then(data => {
+                        console.log("Created new diagram with id: " + data.id + " and title: " + data.title);
+                        diagramId.current = data.id;
+                        setDiagramName(data.title); 
+                        diagramNameRef.current = data.title; //Crutch for the handleSave function
+                        // doExport(); //autosave the new diagram
```
Comment (arthurhohinng) [inline client/src/pages/Editor.tsx:103]: Also here

**Review by arthurhohinng — COMMENTED** [review]
A lot of commented code here, is it intentional for the next PR?
If not, we should clean it up

File: client/src/pages/Editor.tsx (hunk 5 above)
Comment (Vladmidir) [reply client/src/pages/Editor.tsx:220]: Yes

File: client/src/pages/Editor.tsx (hunk 7 above)
Comment (Vladmidir) [reply client/src/pages/Editor.tsx:10]: yes

File: server/models/SolutionInProgress.js (hunk 4 above)
Comment (Vladmidir) [reply server/models/SolutionInProgress.js:4]: Will do

File: client/src/pages/Editor.tsx (hunk 8 above)
Comment (Vladmidir) [reply client/src/pages/Editor.tsx:48]: yes

File: server/index.js [hunk 12]
```diff
@@ -12,27 +12,13 @@ app.use(express.json());
...
-// });
-
 // Sync Sequelize models
 db.sequelize.sync().then(async () => {
   // Use { force: true } cautiously as it will drop existing tables
   console.log("Database synced");
 
   // import the challenges into the db. Comment out after first run
-  // await importChallenges();
+  await importChallenges();
```
Comment (arthurhohinng) [inline server/index.js:21]: intentionally leaving this uncommented?

**Review by arthurhohinng — APPROVED** [review]
Looks good, just one minor comment

File: server/index.js (hunk 12 above)
Comment (Vladmidir) [reply server/index.js:21]: yes. I will delete the the `comment out` part

**Pull request merged**